### PR TITLE
Factory plugin implementation

### PR DESCRIPTION
**Pull request by wjuan-AFK**
Adding implementation in the factory_impl class for loading request source plugins. This also introduces python tests for the previous PR and undoes the temporary hack to reduce test_integration_coverage threshold in issue #564.

### CODE REVIEW

**Comment by dubious90** [comment]
@mum4k Could you please review and merge when done?

File: source/client/factories_impl.h
Comment (mum4k) [inline source/client/factories_impl.h:70]: Can this be private instead or do we have a good reason to make it protected?

File: test/factories_test.cc
Comment (mum4k) [inline test/factories_test.cc:80]: (nit) Can we replace some of these auto statements with the actual type? Only needed in places where the type isn't obvious from the expression.

**Review by mum4k — CHANGES_REQUESTED** [review]
Looks good just one question, let's wait for review from @oschaaf.

File: source/client/factories_impl.h
Comment (wjuan-AFK) [reply source/client/factories_impl.h:70]: I believe that this can be private.

File: source/client/factories_impl.cc
Comment (oschaaf) [inline source/client/factories_impl.cc:183]: Maybe annotate with a `XXX_GCOVR_EXCL_XXX`. We don't care not every branch gets hit in tests, putting that in documents an explicit choice that we're ok with that. If it would ever start working, it would also help with correcting the reported coverage % (it seems to be ignored today).

File: source/client/factories_impl.cc
Comment (oschaaf) [inline source/client/factories_impl.cc:202]: Consider covering this branch in a unit test, or annotating with `XXX_GCOVR_EXCL_XXX` if we don't care.

File: source/client/factories_impl.h
Comment (oschaaf) [inline source/client/factories_impl.h:80]: Not all parameters are documented, consider adding doc comments for `api` & `header`?

File: test/integration/test_request_source_plugin.py
Comment (oschaaf) [inline test/integration/test_request_source_plugin.py:11]: Just a thought, I wonder if it would make sense to parameterize this test over plugin setup. So we iterate the same test over all known plugins. That way, future plugins could add themselves to this hypothetical parameterized list, and have basic integration test coverage through that. (This would imply that all plugin types would be set up with a configuration that yields the same traffic pattern). Wouldn't object to leaving this as is, but if you see sufficient value it might be worth doing.

**Review by oschaaf — COMMENTED** [review]
Looks great, leaving a few small asks, and a few thoughts.

File: test/integration/test_request_source_plugin.py
Comment (wjuan-AFK) [reply test/integration/test_request_source_plugin.py]: Sure thing.

File: source/client/factories_impl.cc
Comment (oschaaf) [reply source/client/factories_impl.cc:183]: As discussed on slack, we could probably just append `// NOT_REACHED_GCOVR_EXCL_LINE` to the line itself, or use 
```c++.
// GCOVR_EXCL_START
... code to be ignored in coverage measurements ...
// GCOVR_EXCL_STOP
```

File: source/client/factories_impl.cc
Comment (oschaaf) [inline source/client/factories_impl.cc:183]: I didn't expect the GCOVR exclusion here, but rather at/around the RELEASE_ASSERT line above -- so, sorry could you remove it here? Looking at the [coverage report](https://22476-180498819-gh.circle-artifacts.com/0/coverage/source/client/factories_impl.cc.gcov.html) everything is fine now, including the RELEASE_ASSERT line, so it seems we don't need it there either now (probably because there's no conditional argument being passed).

File: source/client/factories_impl.cc
Comment (oschaaf) [reply source/client/factories_impl.cc:177]: Thanks for covering this in a test. Resolving.

File: source/client/factories_impl.cc
Comment (oschaaf) [reply source/client/factories_impl.cc:206]: Resolving this, thanks for covering it in a test.

**Review by oschaaf — CHANGES_REQUESTED** [review]
One last nit about removing `// GCOVR_EXCL_START` and `// GCOVR_EXCL_END`. It seems some of the changes outdated that comment, and these annotations no longer have value. Sorry about the back and forth on that. Other then that, this looks great to me.

**Review by mum4k — APPROVED** [review]
Ready for merging, blocked on the clang_tidy flake which we are planning to remove temporarily in the near future (#570).

**Comment by mum4k** [comment]
@oschaaf, this is still marked as "change requested" from your last review, can you please indicate if there is anything outstanding or approve if otherwise?

**Comment by oschaaf** [comment]
@mum4k  Sorry, I forgot to update the `waiting-for-...` label. There's one last (small) [change](https://github.com/envoyproxy/nighthawk/pull/566/files#r520878294) I requested.

**Comment by mum4k** [comment]
Please merge from master to remove the hard block on clang_tidy.

File: source/client/factories_impl.cc
Comment (wjuan-AFK) [reply source/client/factories_impl.cc]: Sounds good.

**Review by oschaaf — APPROVED** [review]
Looks great, thanks

**Review by mum4k — APPROVED** [review]

**Pull request merged**
